torch/csrc/Storage.cpp

```diff
@@ -149,14 +162,34 @@ static bool THPStorage_isPreservable(THPStorage* self) {
 }
 
 static bool THPStorage_tryPreserve(THPStorage* self) {
-  const auto& storage = THPStorage_Unpack(self);
-
   if (!THPStorage_isPreservable(self)) {
     return false;
   }
 
+  const auto& storage = THPStorage_Unpack(self);
   c10::StorageImpl* storage_impl = storage.unsafeGetStorageImpl();
 
+  auto maybe_pyobj = storage_impl->pyobj_slot()->check_pyobj(
+      getPyInterpreter(),
+      /*ignore_hermetic_tls=*/!self->is_hermetic);
+  // NOTE: It is possible to just set the PyObjectSlot here, but the point is
+  // that we should have already set PyObjectSlot when the storage PyObject was
+  // created.
+  TORCH_INTERNAL_ASSERT(
+      maybe_pyobj.has_value(),
+      "Trying to preserve a Python storage whose PyObjectSlot does not have a PyObject");
+
+  PyObject* pyobj = *maybe_pyobj;
+
+  TORCH_CHECK(
+      THPStorage_Check(pyobj),
+      "Expected a storage type, but got ",
+      Py_TYPE(pyobj)->tp_name);
+
+  TORCH_INTERNAL_ASSERT(
+      (void*)pyobj == (void*)self,
+      "Python storage and the PyObject in the internal PyObjectSlot are not at the same address");
```

Comment:
this would have helped with your debugging right? Maybe a good one to add to Tensor too?

Reply:
Actually, I should remove this. It's guaranteed to pass since `THPStorage_isPreservable` already checks whether the storage PyObject matches what's in the PyObjectSlot and returns false if they don't match, so this assert won't be executed in that case. Tensor already does that check as well

Reply:
But the other check above this one checking that the python type is storage is still useful, and I can add it to tensor